inst/include/Rcpp/api/meat/Rcpp_eval.h

```diff
@@ -18,38 +18,95 @@
 #ifndef Rcpp_api_meat_Rcpp_eval_h
 #define Rcpp_api_meat_Rcpp_eval_h
 
+#include <Rcpp/Interrupt.h>
+
 namespace Rcpp{
 
-    inline SEXP Rcpp_eval(SEXP expr_, SEXP env) {
-        Shield<SEXP> expr( expr_) ;
+    struct EvalCall {
+        SEXP expr;
+        SEXP env;
+        SEXP result;
+        std::vector<std::string> warnings;
+        std::string error_message;
+    };
+
+    inline void Rcpp_eval(void* data) {
+
+        EvalCall* evalCall = (EvalCall*)data;
+        SEXP env = evalCall->env;
+
+        Shield<SEXP> expr(evalCall->expr) ;
 
         reset_current_error() ;
 
         Environment RCPP = Environment::Rcpp_namespace();
+        SEXP withCallingHandlersSym    = ::Rf_install("withCallingHandlers");
         SEXP tryCatchSym               = ::Rf_install("tryCatch");
         SEXP evalqSym                  = ::Rf_install("evalq");
         SEXP conditionMessageSym       = ::Rf_install("conditionMessage");
         SEXP errorRecorderSym          = ::Rf_install(".rcpp_error_recorder");
+        SEXP warningRecorderSym        = ::Rf_install(".rcpp_warning_recorder");
+        SEXP collectWarningsSym        = ::Rf_install(".rcpp_collect_warnings");
         SEXP errorSym                  = ::Rf_install("error");
+        SEXP warningSym                = ::Rf_install("warning");
 
-        Shield<SEXP> call( Rf_lang3(
+        // define the tryCatchCall
+        Shield<SEXP> tryCatchCall( Rf_lang3(
             tryCatchSym,
             Rf_lang3( evalqSym, expr, env ),
             errorRecorderSym
         ) ) ;
-        SET_TAG( CDDR(call), errorSym ) ;
-        /* call the tryCatch call */
+        SET_TAG( CDDR(tryCatchCall), errorSym ) ;
+
+        // encose it in withCallingHandlers
+        Shield<SEXP> call( Rf_lang3(
+            withCallingHandlersSym,
+            tryCatchCall,
+            warningRecorderSym
+        ) ) ;
+        SET_TAG( CDDR(call), warningSym ) ;
+
+        // execute the call
         Shield<SEXP> res(::Rf_eval( call, RCPP ) );
 
+        // collect warnings
+        Shield<SEXP> warnings(::Rf_eval(Rf_lang1(collectWarningsSym), RCPP));
```

Comment:
Doing an `eval` on an `Rf_lang` symbol is dangerous -- the language object used for the evaluation can actually be cleaned up and then hence be unavailable for the call, which breaks any functions that rely on `match.call()` or other introspection on the active call.

It's best to create the language object, protect it, evaluate it, then unprotect it after it's not needed.